### PR TITLE
Ayelet k 4

### PR DESCRIPTION
**Pull request by ayeletK**
Hi!
Please review my code :)

1. I am sorry for submitting in the last moment (exams started)
2. I have some TODO in my code - I hope I'll get to it during the iterations
3. I would like to get comments on the tests.cpp - is that how a tests class should look? how can it be improved (I have more edge cases to check)

Thanks!

### CODE REVIEW

File: ayeletK/assignment4/tests.cpp [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+    ~Tests(){
+        if (_delete_data) {
+            for (int row = 0; row < _rows; ++row){
+                delete _data[row];
+            }
+            delete _data;
+        }
+    }
+    void run_tests(){
+        RUN_TEST(test_count_islands_zero_islands);
```
[the remark this answers is not in the document]
Comment (ayeletK) [reply ayeletK/assignment4/tests.cpp]: ok :)

File: ayeletK/assignment4/grid.cpp [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+        _cols = cols_num;
+        for (int row = 0; row < _rows; ++row){
+            _grid.push_back(vector<T>(_cols));
+            for (int col = 0; col < _cols; ++col){
+                _grid[row][col] = data[row][col];
+            }
+        }
+    }
+    explicit Grid(vector<vector<T>>* data): _grid(*data) {}
+    void update_value(Position* pos, T val){
```
[the remark this answers is not in the document]
Comment (ayeletK) [reply ayeletK/assignment4/grid.cpp]: I wanted to save the need of copying the position passed as parameter, and pass the parameter by reference rather then by value.
Is it a waste to do so?

File: ayeletK/assignment4/grid.cpp [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+        if (is_legal_position(pos)) 
+            _grid[pos->get_row()][pos->get_col()] = val;
+    }
+    T get_value(Position* pos){
```
[the remark this answers is not in the document]
Comment (ayeletK) [reply ayeletK/assignment4/grid.cpp]: ok 👍

File: ayeletK/assignment4/grid.cpp [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+        if (is_legal_position(pos)){
+            return _grid[pos->get_row()][pos->get_col()];
+        }
+        return T();
+    }
+    void get_grid_dimentions(int* rows, int* cols){
```
[the remark this answers is not in the document]
Comment (ayeletK) [reply ayeletK/assignment4/grid.cpp]: So in general - if a parameter is not modified, it should be passed as const?

File: ayeletK/assignment4/grid.cpp [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+    }
+    bool is_legal_position(Position* pos){
+        return is_legal_row(pos->get_row()) && is_legal_col(pos->get_col());
+    }
+    unordered_set<Position> get_neighbours(Position* pos, T wanted_val){
+        unordered_set<Position> neighbours;
+        Position neighbour;
+        for (int i: DELTA){
+            for (int j: DELTA){
+                if (abs(i-j) != 1) continue;
```
[the remark this answers is not in the document]
Comment (ayeletK) [reply ayeletK/assignment4/grid.cpp]: I agree

File: ayeletK/assignment4/dfs.cpp [hunk 6]
```diff
@@ -0,0 +1,44 @@
+#include "grid.cpp"
+
+// TODO: should be static class?
```
[the remark this answers is not in the document]
Comment (ayeletK) [reply ayeletK/assignment4/dfs.cpp]: I felt wrong creating new DFS instance every time I wanted to call run_DFS method. That made me think maybe I have need in the method run_DFS and not in the class itself. I mean - it felt like bad practice to create DFS class instance just so I can call run_DFS.

File: ayeletK/assignment4/grid.cpp [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+        return T();
+    }
+    void get_grid_dimentions(int* rows, int* cols){
+        *rows = _rows;
+        *cols = _cols;
+    }
+    bool is_legal_position(Position* pos){
+        return is_legal_row(pos->get_row()) && is_legal_col(pos->get_col());
+    }
+    unordered_set<Position> get_neighbours(Position* pos, T wanted_val){
```
[the remark this answers is not in the document]
Comment (ayeletK) [reply ayeletK/assignment4/grid.cpp]: You are right. I changed the method but not it's name. Will be fixed!

File: ayeletK/assignment4/dfs.cpp [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+template <typename T>
+class DFS{
+    Grid<T>* _grid;
+    T _wanted_val;  // what is the interesting data in the grid.
```
[the remark this answers is not in the document]
Comment (ayeletK) [reply ayeletK/assignment4/dfs.cpp]: I will fix it

File: ayeletK/assignment4/dfs.cpp [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+    
+    unordered_set<Position> get_unvisited_neighbours(Position* curr_pos){
+        unordered_set<Position> res;
+        for (Position pos: _grid->get_neighbours(curr_pos, _wanted_val)){
+            if (_unvisited.find(pos) != _unvisited.end())
+                res.insert(pos);
+        }
+        return res;
+    }
+    void find_DFS_component(Position* curr_pos, unordered_set<Position>* component){
```
[the remark this answers is not in the document]
Comment (ayeletK) [reply ayeletK/assignment4/dfs.cpp]: will be fixed!

File: ayeletK/assignment4/solution.cpp [hunk 10]
```diff
@@ -0,0 +1,34 @@
...
+#include "dfs.cpp"
+#include "grid.cpp"
+typedef Grid<bool> Map;
+
+class Solution{
+    Map _map;
+    DFS<bool> _dfs;
+    
+    // TODO: if I have time - implement iterator for Grid class
+    void collect_all_land_tile_pos(unordered_set<Position>* land_tile){
```
[the remark this answers is not in the document]
Comment (ayeletK) [reply ayeletK/assignment4/solution.cpp]: Ok

File: ayeletK/assignment4/tests.cpp [hunk 11]
```diff
@@ -0,0 +1,87 @@
+#define EXAMPLE_SIZE 4
+#define ONE 1
```
[the remark this answers is not in the document]
Comment (ayeletK) [reply ayeletK/assignment4/tests.cpp]: Should I also drop the EXAMPLE_SIZE definition then?

File: ayeletK/assignment4/tests.cpp [hunk 12]
```diff
@@ -0,0 +1,87 @@
...
+        
+        for (int row = 0; row < _rows; ++row){
+            _data[row] = new bool[EXAMPLE_SIZE];
+            for (int col = 0; col < _cols; ++col){
+                _data[row][col] = example_vec[row][col];
+            }
+        }
+        
+    }
+    bool test_count_islands_zero_islands(){
```
[the remark this answers is not in the document]
Comment (ayeletK) [reply ayeletK/assignment4/tests.cpp]: Got it.

File: ayeletK/assignment4/dfs.cpp [hunk 13]
```diff
@@ -0,0 +1,44 @@
+#include "grid.cpp"
+
+// TODO: should be static class?
+template <typename T>
+class DFS{
```
[the remark this answers is not in the document]
Comment (ayeletK) [reply ayeletK/assignment4/dfs.cpp]: fixed

**Comment by ayeletK** [comment]
I am working on online compiler - the code compiled and the tests ran. The first two passed. The third one I had to debug. I uploaded a new version with corrections of the code according to the review and bug solving. Now it works :)
+ I read the comment you wrote on slack channel about how we are not bound to arrays - if the PL we are working with has better data structures - so I fixed the implementation to work with 2d vector.

File: ayeletK/assignment4/grid.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (Brusi) [reply ayeletK/assignment4/grid.cpp]: In C, you only have two options: Pass by value (which always makes a copy) or pass by pointer (which makes not copy).
In C++, you have more options:
You can pass by value or pointer, but also by reference which does not make a copy, similarly to a pointer.
Also, everything can be either const or non-const.
The default for parameters which you do not need to change, is a const reference - it neither makes a copy each call, and it is safe to not change the value for the caller.

Passing by non-const pointer (or reference) should be reserved for cases in which you want the function to modify the value for the caller (i.e. output parameter, or input+output parameter).

Examples:
```C++
// Fills the position of val in the grid. Does not change the grid,
// and writes position into i and j.
void FindInGrid(const Grid& grid, const T& val, int* i, int* j);  

// Here grid is modified by the function so it is passed as a non-const reference.
void SetInGrid(Grid& grid, const T& val, int i, int j);
```

File: ayeletK/assignment4/grid.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (Brusi) [reply ayeletK/assignment4/grid.cpp]: Yes :)

File: ayeletK/assignment4/dfs.cpp (hunk 6 above)
[the remark this answers is not in the document]
Comment (Brusi) [reply ayeletK/assignment4/dfs.cpp]: It is usually a good idea to encalsulate some logic inside a class.
I liked that the grid is a "dumb" class - it only holds values and nothing more.
Any other calculations (e.g. DFS, finding components) happens in a dedicated class. I think this is a good practice.

File: ayeletK/assignment4/tests.cpp (hunk 11 above)
[the remark this answers is not in the document]
Comment (Brusi) [reply ayeletK/assignment4/tests.cpp]: It seems now that you do not use it anymore, so you can drop it :)

File: ayeletK/assignment4/grid.cpp [hunk 14]
```diff
@@ -3,59 +3,52 @@
...
-            for (int j: DELTA){
-                if (abs(i-j) != 1) continue;
-                neighbour = Position(pos->get_row()+i, pos->get_col()+j);
-                if (is_legal_position(&neighbour) && get_value(&neighbour) == wanted_val)
-                        neighbours.insert(neighbour);
-            }
+        int row, col;
+        for (Position pos: delta_positions){
+            neighbour = curr_pos + pos;
+            if (neighbour == curr_pos) continue;
```
Comment (Brusi) [inline ayeletK/assignment4/grid.cpp]: Why would the neighbor be the same as the current position? Does this solve a real problem or is it just for safety?

Industry secret:
If that's just for safety (and you do not expect it to actually happen in runtime), you can replace this with an assert clause:
`assert(neighbor != current_pos);`
This way if this bug happens then this will not fail silently but crash the program so you can find it and debug it.

If you are not worried that it will actually happen, it is best to remove it completely to make the code cleaner.
